packages/server/src/auth/types/request.ts

```diff
@@ -11,15 +12,101 @@ export function isFetchRequest(request: GenericRequest): request is Request {
     return "headers" in request && typeof request.headers.get === "function";
 }
 
-export function isNodeResponse(response: GenericResponse): response is ServerResponse {
-    return "setHeader" in response;
+interface RequestAdapter {
+    getCookieHeader: () => string | null | undefined;
+    getBody: () => Promise<any>;
 }
 
-export function isFetchResponse(response: GenericResponse): response is Response {
-    return "headers" in response && typeof response.headers.append === "function";
+export class NodeRequestAdapter implements RequestAdapter {
+    constructor(private request: IncomingMessage) {}
+
+    getCookieHeader() {
+        return this.request.headers.cookie;
+    }
+
+    getBody() {
+        return getNodeRequestBody(this.request);
+    }
+}
+
+export class FetchRequestAdapter implements RequestAdapter {
+    constructor(private request: Request) {}
+
+    getCookieHeader() {
+        return this.request.headers.get("Cookie");
+    }
+
+    getBody() {
+        return this.request.json();
+    }
+}
+
+export function isNodeResponse(response?: GenericResponse): response is ServerResponse {
+    return response != null && "setHeader" in response;
+}
+
+export function isFetchResponse(response?: GenericResponse): response is Response {
+    return response != null && "headers" in response && typeof response.headers.append === "function";
+}
+
+interface ResponseAdapter {
+    setCookies: (cookies: string[]) => void;
+    setAuthMaterial: (authMaterial: AuthMaterialWithUser) => GenericResponse | null;
+    setError: (statusCode: number, errorResponseBody: ErrorBody) => GenericResponse;
+}
+
+export class NodeResponseAdapter implements ResponseAdapter {
+    constructor(private response?: ServerResponse) {}
+
+    setCookies(cookies: string[]) {
+        if (this.response == null) {
+            throw new Error("Response not found");
+        }
+        this.response.setHeader("Set-Cookie", cookies);
+    }
+
+    setAuthMaterial(authMaterial: AuthMaterialWithUser) {
+        if (this.response == null) {
+            return null;
+        }
+        this.response.setHeader("Content-Type", "application/json");
+        this.response.write(JSON.stringify(authMaterial));
+        return this.response;
+    }
+
+    setError(statusCode: number, errorResponseBody: ErrorBody) {
+        if (this.response == null) {
+            throw new Error("Response not found");
+        }
+        return setNodeResponseError(this.response, statusCode, errorResponseBody);
+    }
+}
+
+export class FetchResponseAdapter implements ResponseAdapter {
+    constructor(private response?: Response) {}
+
+    setCookies(cookies: string[]) {
+        if (this.response == null) {
+            throw new Error("Response not found");
+        }
+
+        cookies.forEach((cookie) => this.response?.headers.append("Set-Cookie", cookie));
+    }
+
+    setAuthMaterial(authMaterial: AuthMaterialWithUser) {
+        return new Response(JSON.stringify(authMaterial), {
```

Comment:
Isn't this method loosing state set to `this.response` before? i.e. cookies?

Reply:
i create a different adapter for each instance. for cookies, I need to pass in a response, for `handleCustomRefresh` I need to create the response instance